src/utils/retry.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/example-shopping
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {HttpErrors} from '@loopback/rest';
+import {promisify} from 'util';
+import * as debugFactory from 'debug';
+const debug = debugFactory('loopback:example:shopping');
+
+export interface TaskStatus<T> {
+  done: boolean;
+  value?: T | null;
+}
+
+/**
+ * A task that can be retried
+ */
+export interface Task<T> {
+  run(): Promise<TaskStatus<T>>;
+  description: string;
+}
+
+/**
+ * Options for retry
+ */
+export interface RetryOptions {
+  /**
+   * Maximum number of tries including the first run.
+   */
+  maxTries?: number;
+  /**
+   * Milliseconds to wait after each try
+   */
+  interval?: number;
+}
+
+/**
+ * Retry a task for number of times with the given interval in ms
+ * @param task Task object {run, description}
+ * @param maxTries Maximum number of tries (including the first run),
+ * default to 10
+ * @param interval Milliseconds to wait after each try, default to 100ms
+ */
+export async function retry<T>(
+  task: Task<T>,
+  {maxTries = 10, interval = 100}: RetryOptions = {},
+): Promise<T> {
+  if (maxTries < 1) maxTries = 1;
+  let triesLeft = maxTries;
+  while (true) {
+    debug(
+      'Try %s (%d/%d)',
+      task.description,
+      maxTries - triesLeft + 1,
+      maxTries,
+    );
+    const status = await task.run();
+    if (status.done) return status.value!;
+    if (--triesLeft > 0) {
+      debug('Wait for %d ms', interval);
+      await sleep(interval);
+    } else {
+      // No more retries, timeout
+      const msg = `Failed to ${task.description} after ${maxTries *
+        interval} ms`;
+      debug('%s', msg);
+      throw new HttpErrors.RequestTimeout(msg);
```

Comment:
Nitpick: Since `retry` is not used at REST layer, but inside a repository, I think we shouldn't be throwing an HttpError and use `err.code` to provide a machine-readable flag that can be used to detect this particular error.